examples/typescript-example/src/App.tsx

```diff
@@ -0,0 +1,226 @@
+  extends ColumnInstance<D>,
+    UseSortByColumnProps<D>,
+    UseResizeColumnsHeaderProps<D>,
+    UseFiltersColumnProps<D> {}
+
+const Rigth = styled.div`
+  text-align: right;
+`
+
+const OrderIndicator = styled.span<{ isSorted?: boolean; isSortedDesc?: boolean }>`
+  position: absolute;
+  right: 0;
+  left: 0;
+  height: 2px;
+  background: black;
+  ${({ isSorted, isSortedDesc }) => (isSorted ? (isSortedDesc ? { top: 0 } : { bottom: 0 }) : { display: 'none' })}
+`
+
+const TableWrap = styled.div`
+  display: inline-block;
+  border-spacing: 0;
+  border: 1px solid black;
+`
+
+const TableHead = styled.div``
+const TableBody = styled.div``
+const TableRow = styled.div``
+
+const TableCell = styled.div`
+  position: relative;
+  margin: 0;
+  padding: 0.5rem;
+  border-bottom: 1px solid black;
+  border-right: 1px solid black;
+  white-space: nowrap;
+  overflow: hidden;
+  text-overflow: ellipsis;
+  :last-child {
+    border-right: 0;
+  }
+`
+const Resizer = styled.div`
+  display: inline-block;
+  width: 5px;
+  height: 100%;
+  position: absolute;
+  right: 0;
+  top: 0;
+  transform: translateX(50%);
+  z-index: 1;
+`
+
+const Input = styled.input`
+  width: calc(100% - 0.5rem);
+`
+
+const useStopPropagationOnClick = <T extends any>(ref: T) => {
+  useEffect(() => {
+    if (ref.current !== null) {
+      ref.current.addEventListener('click', (e: MouseEvent) => e.stopPropagation())
+    }
+  }, [])
+}
+
+const DefaultColumnFilter = ({
+  column: { filterValue, preFilteredRows, setFilter }
+}: {
+  column: TableColumn<Data>
+}) => {
+  const count = preFilteredRows.length
+  const inputRef = useRef<HTMLInputElement>(null)
+  useStopPropagationOnClick(inputRef)
+  return (
+    <Input
+      ref={inputRef}
+      value={filterValue || ''}
+      onChange={e => {
+        setFilter(e.target.value || undefined) // Set undefined to remove the filter entirely
+      }}
+      placeholder={`Search ${count} records...`}
+    />
+  )
+}
+
+const ResizerComponent: FC = props => {
+  const ref = useRef<HTMLDivElement>(null)
+  useStopPropagationOnClick(ref)
+  return <Resizer {...props} ref={ref} />
+}
+
+const defaultColumn = {
+  minWidth: 20,
+  width: 150,
+  maxWidth: 500,
+  Filter: DefaultColumnFilter
+}
+
+const Table: FC<Props> = ({ columns, data }) => {
+  // Use the state and functions returned from useTable to build your UI
+  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } = useTable<Data>(
+    {
+      columns,
+      defaultColumn,
+      data
+    },
+    useFilters,
+    useSortBy,
+    useBlockLayout,
+    useResizeColumns
+  ) as TableInstance<object>
+  // Render the UI for your table
+  return (
+    <TableWrap {...getTableProps()}>
+      <TableHead>
+        {headerGroups.map(headerGroup => (
+          <TableRow {...headerGroup.getHeaderGroupProps()}>
+            {headerGroup.headers.map(c => {
+              const column = (c as unknown) as TableColumn<Data>
```

Comment:
@ggascoigne what do you think about this part, it doesn't feel very good to cast it this way but at least it works, I didn't add the global file but the point is add the type where you need them, I think this will work just fine, just is not super pretty

Reply:
headerGroup.headers is a ColumnInstance[] (ok, I've not checked this, but I'm pretty sure that it is).  This is why I use dependency merging rather then defining new types.

The types for all of these objects and interfaces have to be defined and you can't reasonably extend them into new types, but you can extend the existing types.

There's no easy way to have headerGroup.headers reflect a new type.

But, if you follow the pattern in react-table-config.d.ts, you can extend the definition of ColumnInstance so that it simply returns the right data in the existing interface.

I completely agree that this is a bit wonky, I'm not sure how to make it better though, there's no getting around the global nature of the definitions.  At least, we haven't found one yet.

Reply:
But that's why I would rather cast to the correct tipe when I need it instead of creating a global types that it will be a mess as soon as you have two different table. I don't know either if there is a better solution for this, I will try to come up with something better perhaps

Reply:
I'm working on a new approach for instance level configuration of the types. In the meantime, I agree with @ggascoigne that interface merging is the best approach. Since the interface merged types are a superset of all tables it will work for all whereas casting can erroneously cast to whatever you want.

Reply:
agree on that, but we need a way to extend the column instance (as far as I see it's the most problematic one) from the use table function to make it the cleanest, not sure it's possible.

Reply:
Are there any example yet where interface merging is used? The typing work that's been done is great, but it's not obvious how to consume this yet IMO. @dbertella example is great and I think many (e.g. me!) will benefit with having an approach on how to use react-table with TypeScript.

Reply:
take a look at https://github.com/ggascoigne/react-table/tree/ts-pagination-example/examples/pagination-ts - there's something not quite right about the codesandbox config for this, but if you actually check it out and compile it, it compiles fine and vscode correctly displays the type information where you'd want it to.  For some reason codesandbox highlights errors that aren't there - need to work that out.

Reply:
@dbertella what about
```
interface SortableColumnInstance<D extends object = {}>
  extends ColumnInstance<D>,
  UseSortByColumnProps<D> {}  
```
```
        {headerGroup.headers.map((column: SortableColumnInstance) => (
```

Reply:
@dcchuck I might have missed this comment, sorry that looks good if it works, but I guess nowadays everybody is doing this merging on the global types isn't it?